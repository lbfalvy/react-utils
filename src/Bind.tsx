import React, { SyntheticEvent } from "react";

type Props<T, Key extends keyof T, U> = Omit<U, 'value' | 'onChange'> & { 
    object: T
    index: Key
    for: React.ComponentType<{ 
        value?: any,
        onChange?: ((value: any) => any) | undefined
    } & U>
    dom?: boolean
}

/**
 * Link an input or other component to an entry on an object. The input
 * will own the entry, and capping or setting the value on the receiving
 * end will not be possible.
 */
export function Bind<T, K extends keyof T, U>(props: Props<T, K, U>): React.ReactElement {
    const {for: Component, object, index, dom, ...restProps} = props;
    const [value, setValue] = React.useState(object[index]);
    React.useLayoutEffect(() => { object[index] = value; }, [object, index, value]);
    return <Component value={value} onChange={(ev: any) => {
        if (!dom) return setValue(ev)
        setValue(ev.target?.value)
    }} {...restProps as unknown as U} />;
}