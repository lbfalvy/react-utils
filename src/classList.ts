/**
 * Decide whether each class gets applied with a simple boolean expression
 * @param classes An array of strings, arrays of strings.
 * @returns A composed class list string
 */
export function classList(...classes: any[]): string {
    return classes.flat().filter(c => typeof c == 'string').join(' ')
}