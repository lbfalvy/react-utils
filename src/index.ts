export { Bind } from './Bind'
export { classList } from './classList'
export { mergeRefs } from './mergeRefs'
export { useArray } from './useArray'
export { useChangingHandle } from './useChangingHandle'
export { usePointer } from './usePointer'
export { useScripts } from './useScripts'
export { useWindowDimensions } from './useWindowDimensions'
export { useVariable } from "./useVariable";