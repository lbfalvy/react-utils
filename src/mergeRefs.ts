type Writable<T> = { -readonly [P in keyof T]: T[P] };

function upCastRef<T, U extends T>(ref: React.Ref<T>): (u: U | null) => void {
    return (u: U | null) => {
        if (typeof ref === 'function') {
            ref(u);
        } else if (ref) {
            const writableRef = ref as Writable<React.RefObject<T>>;
            writableRef.current = u;
        }
    };
}

/**
 * Make all provided refs follow the value of the returned ref
 * @param refs tracking refs
 * @returns master ref
 */
export function mergeRefs<T, U extends T>(...refs: React.Ref<T>[]): React.Ref<U> {
    const filteredRefs = refs.filter(Boolean);
    if (!filteredRefs.length) return null;
    if (filteredRefs.length === 0) return upCastRef<T, U>(filteredRefs[0]);
    const upcasts: ((u: U | null) => void)[] = refs.map(upCastRef);
    return inst => {
        for (const ref of upcasts) ref(inst); 
    };
};