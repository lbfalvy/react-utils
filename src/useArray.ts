import React from "react";

/**
 * Ensures referential equality of an array without changed
 * dependency array length warnings.
 * @param input Array
 * @returns Array but memo'd based on its entries.
 */
export function useArray<T extends unknown[]>(input: T): T {
    const ref = React.useRef<T>(input)
    const cur = ref.current
    if (input.length === cur.length
        && input.every((v, i) => v === cur[i]))
        return cur
    return ref.current = input
}