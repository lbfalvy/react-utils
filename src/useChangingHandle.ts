import React from "react";

/**
 * useImperativeHandle but it changes when you say so.
 * @param ref Incoming ref object or callback
 * @returns Function to set the ref value
 */
export function useChangingHandle<T>(ref?: React.Ref<T>): null | undefined | ((value: T) => void) {
    return React.useMemo(
        () => typeof ref == 'function' || !ref ? ref
            : value => { (ref as React.MutableRefObject<T>).current = value },
        [ref]
    )
}