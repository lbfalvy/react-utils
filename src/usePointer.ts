import React from "react";

function setPosition(ev: MouseEvent) {
    position.now = Date.now();
    position.x = ev.clientX;
    position.y = ev.clientY;
}

interface MousePosition {
    x: number
    y: number
    now: number
}

const position = { x: 0, y: 0, now: 0 };
let watchers = 0;

/**
 * Keep track of the mouse pointer, ref style.
 * @returns A function that always returns the corrent mouse position
 */
export function usePointer(handler: (p: MousePosition) => any): () => MousePosition {
    React.useEffect(() => {
        if (watchers++ == 0) window.addEventListener('mousemove', setPosition);
        return () => {
            if (--watchers == 0) window.removeEventListener('mousemove', setPosition);
        };
    }, []);
    React.useEffect(() => {
        if (!handler) return;
        const cb = () => handler({ ...position })
        window.addEventListener('mousemove', cb)
        return () => window.removeEventListener('mousemove', cb)
    }, [handler])
    return () => ({ ...position });
}