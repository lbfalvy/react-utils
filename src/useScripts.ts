import React from "react";
import { flat, allSettled, XPromise, resolve } from '@lbfalvy/when'
import { useArray } from "./useArray";

export type SettledResult =
    | { status: 'fulfilled', value: string }
    | { status: 'rejected', reason: string }
    | { status: 'cancelled' }

export function useScripts(source: string[]): XPromise<SettledResult[]> {
    const scripts = React.useRef<HTMLScriptElement[]>([]);
    source = useArray(source)
    React.useEffect(() => () => {
        for (const script of scripts.current) script.remove();
    }, [])
    return React.useMemo(() => {
        for (const script of scripts.current) if (!source.includes(script.src)) script.remove();
        const newScripts: HTMLScriptElement[] = [];
        if (0 < source.length) {
            const onloads: XPromise<string>[] = [];
            const parent = document.getElementById('scripts') ?? document.createElement('div');
            parent.id = 'scripts';
            parent.style.display = 'none';
            // New scripts
            for (const url of source) if (!scripts.current.find(el => el.src == url)) {
                const script = document.createElement('script');
                script.src = url;
                const [promise, resolve, reject] = flat<string>();
                script.onload = () => resolve(url);
                script.onerror = () => reject(url);
                parent.appendChild(script);
                newScripts.push(script);
                onloads.push(promise);
            }
            if (!parent.parentElement) document.body.appendChild(parent);
            scripts.current = newScripts;
            return allSettled(onloads);
        } else {
            scripts.current = [];
            const elem = document.getElementById('scripts')
            if (elem?.childNodes.length == 0) elem.remove();
            return resolve<[]>([]);
        }
    }, [source]);
}