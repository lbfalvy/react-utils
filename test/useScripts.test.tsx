import { timeout, XPromise } from '@lbfalvy/when'
import { JSDOM, DOMWindow } from 'jsdom'
import React from 'react'
import { create, ReactTestRenderer } from 'react-test-renderer'
import { SettledResult, useScripts } from '../src/useScripts'

function Mock({ s, cb }: { s: string[], cb: (p: XPromise<SettledResult[]>) => any }) {
    const p = useScripts(s)
    React.useLayoutEffect(() => cb(p), [cb, p])
    return null
}

let dom: JSDOM
let document: DOMWindow['document']
let cb: jest.Mock<void, [XPromise<SettledResult[]>]>
let component: ReactTestRenderer

beforeEach(() => {
    dom = new JSDOM()
    // @ts-expect-error
    global.window = dom.window
    global.document = document = dom.window.document
    cb = jest.fn()
    component = create(<Mock s={['a', 'b']} cb={cb} />)
})

test('creates and populates script div', () => {
    const scripts = document.getElementById('scripts')
    expect(scripts?.outerHTML).toMatchSnapshot()
})

test('removes script', () => {
    component.update(<Mock s={['a']} cb={cb} />)
})

test('inserts new script without touching later one', () => {
    const b = document.querySelector('[src="b"]')
    expect(b).not.toBeNull()
    component.update(<Mock s={['a', 'c', 'b']} cb={cb} />)
    expect(document.getElementById('scripts')?.outerHTML).toMatchSnapshot()
    expect(document.querySelector('[src="b"]')).toBe(b)
})

test('calls callback when every script settled', async () => {
    const promise: XPromise<any> = cb.mock.calls[0][0]
    promise.execute()
    expect(promise.status).toBe('pending')
    document.querySelector('[src="a"]')?.dispatchEvent(new dom.window.CustomEvent('error'))
    expect(promise.status).toBe('pending')
    document.querySelector('[src="b"]')?.dispatchEvent(new dom.window.CustomEvent('load'))
    await timeout(0)
    expect(promise.value).toHaveLength(2)
    expect(promise.value).toContainEqual({ status: 'fulfilled', value: 'b' })
    expect(promise.value).toContainEqual({ status: 'rejected', reason: 'a' })
})