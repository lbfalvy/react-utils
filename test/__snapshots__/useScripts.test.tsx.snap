// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates and populates script div 1`] = `"<div id=\"scripts\" style=\"display: none;\"><script src=\"a\"></script><script src=\"b\"></script></div>"`;

exports[`inserts new script without touching later one 1`] = `"<div id=\"scripts\" style=\"display: none;\"><script src=\"a\"></script><script src=\"b\"></script><script src=\"c\"></script></div>"`;
